CAL-215 Adds security audit plugin

vinamartin opened the pull request:
#### What does this PR do?
Adds a security audit plugin with a configuration for security attributes that should be audited when changed.

Looking for feedback on the audit message, specifically, whether or not it should say something else and in addition if there seems there is anything missing from the configuration file. 

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?

@codice/security 

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@lessarderic 
@stustison

#### How should this be tested?
Ingest anything into the catalog and then change a security attribute (this is easier with the Catalog UI in DDF). Verify the change was audited. 

#### What are the relevant tickets?
[CAL-215](https://codice.atlassian.net/browse/CAL-215)
#### Checklist:
- [ ] Documentation Updated
	- [ ] Change Log Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

vinamartin commented on catalog/security/security-app/pom.xml:
```diff
@@ -53,8 +55,12 @@
         </dependency>
+        <dependency>
+            <groupId>org.codice.alliance.catalog.plugin</groupId>
+            <artifactId>catalog-plugin-security-audit</artifactId>
```
This seems to be a duplicate? I'll remove this.

bdeining commented on catalog/plugin/catalog-plugin-security-audit/pom.xml:
```diff
@@ -0,0 +1,116 @@
+                                            <counter>COMPLEXITY</counter>
+                                            <value>COVEREDRATIO</value>
+                                            <minimum>0.46</minimum>
+                                        </limit>
```
Why aren't `INSTRUCTION` and `BRANCH` included for Jacoco?

vinamartin replied:
Good question, I thought we were generally okay with excluding them if they were above the threshold, but I can add them back in for clarity here.

bdeining commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,121 @@
+    @Override
+    public UpdateRequest processPreUpdate(UpdateRequest updateRequest,
+            Map<String, Metacard> existingMetacards) throws StopProcessingException {
+
```
What if updateRequest is null?

clockard replied:
Probably want to have a check to see if the request is a remote update. Don't know that we need to update the local log if the change isn't local. you should be able to use `Requests.isLocal(updateRequest)`

vinamartin replied:
Good point, I'll add a null check here.

vinamartin replied:
Added

bdeining commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,121 @@
+                        Attribute oldValue = existingMetacard.getAttribute(descriptorName);
+                        if (!oldValue.equals(newValue)) {
+                            SecurityLogger.audit(
+                                    descriptor.getName() + " with value " + oldValue.getValue()
```
Can use String.format here

clockard replied:
could also switch the the parameterized method

vinamartin replied:
Will change this to use that.

clockard commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,121 @@
+
+        for (Map.Entry<Serializable, Metacard> entry : updateRequest.getUpdates()) {
+            Metacard updateMetacard = entry.getValue();
+            Metacard existingMetacard = existingMetacards.get(entry.getKey()
```
I think it is possible for `existingMetacards` to be empty. One of the access plugins checks for this and will throw a stop processing request but you plugin might run before that check happens.

vinamartin replied:
I'll give it a check.

clockard commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,121 @@
+                    Attribute newValue = updateMetacard.getAttribute(descriptorName);
+                    if (existingMetacard.getAttribute(descriptorName) != null) {
+                        Attribute oldValue = existingMetacard.getAttribute(descriptorName);
+                        if (!oldValue.equals(newValue)) {
```
Are we guaranteed that Attributes with the same values will equal each other? The ddf implementation does but there is no requirement to use it. 
Might want to compare the serialized values instead.

vinamartin replied:
Updated.

clockard commented on catalog/plugin/catalog-plugin-security-audit/src/main/resources/OSGI-INF/blueprint/blueprint.xml:
```diff
@@ -0,0 +1,50 @@
+                <value>security.other-dissemination-controls</value>
+                <value>security.owner-producer</value>
+                <value>security.releasability</value>
+                <value>ext.metadata-originator-classification</value>
```
I don't think we can put these `ext.metadata*` values in ddf. The other upstream projects would need to overwrite the configuration to add these. @stustison what do you think.

clockard commented on catalog/plugin/catalog-plugin-security-audit/src/main/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -0,0 +1,34 @@
+            required="true"
+            type="String"
+            cardinality="1000"
+            default="security.access-groups,security.access-individuals,security.classification,security.classification-system,security.codewords,security.dissemination-controls,security.other-dissemination-controls,security.owner-producer,security.releasability,ext.metadata-originator-classification,ext.metadata-classification,ext.metadata-classification-system,ext.metadata-dissemination-control ...
```
Same comment about the `ext.metadata` values

bdeining commented on catalog/plugin/catalog-plugin-security-audit/src/main/resources/OSGI-INF/blueprint/blueprint.xml:
```diff
@@ -0,0 +1,50 @@
+        </property>
+    </bean>
+
+    <service ranking="99" ref="securityAuditPlugin" interface="ddf.catalog.plugin.AccessPlugin"/>
```
Is there a reason for giving a service ranking of 99?

vinamartin replied:
I'm not sure. Any suggestions here?

bdeining replied:
If the order doesn't matter, don't think it will need a ranking; UpdateOperations just iterates through all registered AccessPlugins

https://github.com/codice/ddf/blob/8af57f314b111a5a42bcce553f531d2339b15217/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/UpdateOperations.java#L551

vinamartin replied:
Fair enough, I don't think order matters here so I will go ahead and update this.

bdeining commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,121 @@
+
+    private List<String> auditAttributes = new ArrayList<>();
+
+    public SecurityAuditPlugin(List<String> auditAttributes) {
```
Is this constructor necessary?  I see that `auditAttributes` is a property on the bean

vinamartin replied:
I was using this for the tests. I'm open to suggestions here though.

bdeining replied:
You could use `setAuditAttributes` in the unit tests and remove this constructor

bdeining commented on catalog/plugin/catalog-plugin-security-audit/src/test/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPluginTest.java:
```diff
@@ -0,0 +1,165 @@
+                updateMetacard));
+
+        Map<String, Metacard> existingMetacards = new HashMap<>();
+        existingMetacards.put("metacardKey", existingMetacard);
```
`metacardKey` can be made into a constant

vinamartin replied:
Will update this.

clockard commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,129 @@
+    @Override
+    public UpdateRequest processPreUpdate(UpdateRequest updateRequest,
+            Map<String, Metacard> existingMetacards) throws StopProcessingException {
+        if (updateRequest != null && Requests.isLocal(updateRequest)) {
```
Could we change this if to do the opposite check and then do a quick return to avoid so much nesting. Could also add the `MapUtils.isNotEmpty(existingMetacards)` check to this if and remove two levels of nesting.

vinamartin replied:
I can do that. The only reason why I had it like this was to make it less weird with the two exactly identical returns.

clockard commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,129 @@
+                            if (updateMetacard.getAttribute(descriptorName) != null) {
+                                Serializable newValue = updateMetacard.getAttribute(descriptorName)
+                                        .getValue();
+                                if (existingMetacard.getAttribute(descriptorName) != null) {
```
Since getAttribute is a fast map lookup we could just get both the existing and update attribute and then check to see if they are null in the same if statement as the `auditAttributes.contains(descriptorName)` and then do a `continue` if the descriptorName is not an audit attribute or one of the metacards doesn't have the attribute.

vinamartin replied:
Sure. Makes sense!

clockard commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,129 @@
+                    for (AttributeDescriptor descriptor : attributeDescriptors) {
+                        String descriptorName = descriptor.getName();
+                        if (auditAttributes.contains(descriptorName)) {
+                            if (updateMetacard.getAttribute(descriptorName) != null) {
```
What if the existing metacard has one of these auditable attributes but the update removes that attribute? Do we not want to log that as well?

vinamartin replied:
Ah forgot that was a case. I can edit this so if the update metacard doesn't have that attribute we should log it as well

lessarderic commented on catalog/plugin/catalog-plugin-security-audit/src/main/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -0,0 +1,34 @@
+    <OCD name="Security Audit Plugin"
+         id="org.codice.alliance.catalog.plugin.security.audit.SecurityAuditPlugin"
+         description="Plugin to audit security attribute changes">
+        <AD name="Security Attributes to audit" id="auditAttributes"
```
Nit: lowercase _a_ for _Security attributes_?

vinamartin replied:
Changed!

lessarderic commented on catalog/plugin/catalog-plugin-security-audit/src/test/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPluginTest.java:
```diff
@@ -0,0 +1,190 @@
+                "A test string",
+                "A test string"));
+    }
+}
```
We should have tests for `null` attributes and non-local ones.

vinamartin replied:
Will add these.

vinamartin commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,134 @@
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+
+package org.codice.alliance.catalog.plugin.security.audit;
```
Pinging @stustison for this question

lessarderic replied:
Still waiting on @stustison?

stustison replied:
No, I don't think there is any specific reason why this plugin needs to be in Alliance. However, if we move it out of here, we will need to still drop a configuration for Alliance to hit all of the additional security attributes.

stustison replied:
@lessarderic @vinamartin A ticket sounds good to me!

vinamartin commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,134 @@
+import ddf.catalog.plugin.StopProcessingException;
+import ddf.security.common.audit.SecurityLogger;
+
+public class SecurityAuditPlugin implements AccessPlugin {
```
Yes, we should, I'll add it.

vinamartin commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,134 @@
+                String descriptorName = descriptor.getName();
+                if (auditAttributes.contains(descriptorName) && existingMetacard.getAttribute(
+                        descriptorName) != null) {
+                    Serializable newValue;
```
Yeah that actually makes a lot of sense. I can refactor this

vinamartin commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,134 @@
+                        newValue = updateMetacard.getAttribute(descriptorName)
+                                .getValue();
+                    } else {
+                        newValue = StringUtils.EMPTY_STRING;
```
Yeah, looking back on this, I am kind of confused myself. I think what I was trying to achieve here is the case when the attribute has been deleted.

vinamartin replied:
Yeah, looking back on this, I am kind of confused myself. I think what I was trying to achieve here is the case when the attribute has been deleted.

vinamartin commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,134 @@
+                            .getValue();
+                    if (!oldValue.equals(newValue)) {
+                        SecurityLogger.audit(String.format(
+                                "%s attribute on metacard %s with value %s was updated to value %s",
```
Sounds good to me

vinamartin commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    public static boolean isLocal(Request req) {
+        return req == null || !req.hasProperties() || isLocal((Request) req.getProperties());
```
I'm not sure....I think this is a case of a bad copy/paste. I need to look through this again and figure out what happened.

vinamartin replied:
Fixed this. For some reason I was missing the second part. :)

vinamartin commented:
@lessarderic Addressed some of your comments. Thanks for the feedback! 

Was wondering what you had in mind as far as itests go and where the docs should live?

brianfelix commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -61,26 +68,30 @@ public UpdateRequest processPreUpdate(UpdateRequest updateRequest,
+                        if (updateMetacard.getAttribute(descriptorName) == null ||
+                                updateMetacard.getAttribute(descriptorName)
+                                        .getValues() == null) {
+                            updateValue = "[DELETED]";
```
Consider making this a constant since it is used in more than one place

vinamartin replied:
Good call!

brianfelix commented on catalog/plugin/catalog-plugin-security-audit/src/main/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -17,7 +17,7 @@
          id="org.codice.alliance.catalog.plugin.security.audit.SecurityAuditPlugin"
          description="Plugin to audit security attribute changes">
-        <AD name="Security Attributes to audit" id="auditAttributes"
+        <AD name="Security attributes to audit" id="auditAttributes"
```
Should this be changed?

vinamartin replied:
Yeah was addressing Eric's comment. :)

lessarderic commented on catalog/plugin/catalog-plugin-security-audit/src/main/java/org/codice/alliance/catalog/plugin/security/audit/SecurityAuditPlugin.java:
```diff
@@ -0,0 +1,164 @@
+
+    private boolean hasAttributeValueChanged(Attribute oldAttribute, Attribute newAttribute) {
+        // The old attribute is null, so it can't be "updated"
+        if (oldAttribute == null || CollectionUtils.isEmpty(oldAttribute.getValues())) {
```
Should this method return `true` if the attribute didn't exist (`oldAttribute == null`) but was added (`newAttribute != null`)?

vinamartin replied:
Oh oops, looks like I forgot that case. I'll update this.

vinamartin commented:
Still actively working through changes on this

vinamartin commented:
Looking for additional feedback now.

lessarderic commented:
@vinamartin, I thought you added documentation and integration tests for this but I don't see any changes related to those in the PR.

vinamartin commented:
@lessarderic I've temporarily undone the itest changes for the sake of review as I'm still struggling with them. As for documentation, was waiting for some more feedback, but I'm starting on it now.

lessarderic commented:
@vinamartin: That works. Let me know when those are in and I'll take another look. Everything else looked good to me.

vinamartin commented:
Pinging @ricklarsen to check the documentation I've added.

ricklarsen commented on distribution/docs/src/main/resources/_contents/_plugins/security-audit-plugin-contents.adoc:
```diff
@@ -0,0 +1,19 @@
+To configure the Catalog Backup Plugin:
+
+. Navigate to the *${admin-console}*.
+. Select ${alliance-security} application.
```
can you bold the app name too?

vinamartin replied:
Gotcha can do

ricklarsen commented on distribution/docs/src/main/resources/draft-documentation.adoc:
```diff
@@ -574,6 +574,8 @@ include::{adoc-include}/_plugins/access-plugin-intro-contents.adoc[]
 
 include::{adoc-include}/_plugins/operation-plugin-contents.adoc[]
+
+include::{adoc-include}/_plugins/security-audit-plugin-contents.adoc[]
```
the `{adoc-include}` will need to be `{adoc-include-cal}` to find files in Alliance.

vinamartin replied:
Ahhh okay, makes sense

lessarderic commented on distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/SecurityAuditPluginTest.java:
```diff
@@ -0,0 +1,97 @@
+        Dictionary properties = new Hashtable<>();
+        properties.put("auditAttributes", attributes);
+        config.update(properties);
+        getServiceManager().waitForAllBundles();
```
We shouldn't have to call `waitForAllBundles()` here. Instead, we should wait for the log message that says that the configuration has been updated and then move on.

lessarderic commented on distribution/test/itests/test-itests-alliance/src/test/java/org/codice/alliance/test/itests/SecurityAuditPluginTest.java:
```diff
@@ -0,0 +1,97 @@
+
+        String log = getFileFromAbsolutePath(
+                System.getProperty("karaf.data") + "/log/security.log");
+        assertTrue(log.contains(String.format(auditMessageFormat,
```
We should use another wait here just in case it takes some time for the log to be added to the file.

bdeining replied:
Also, we should be using `assertThat` vs `assertTrue`
https://codice.atlassian.net/wiki/display/DDF/Unit+Testing

vinamartin replied:
Good point. Once I add in the wait, this assertion will probably disappear.
